File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        Queue<Integer> q = new ArrayDeque<>();
        int[] visit = new int[words.length];
        for(int i=0; i<words.length; i++){
            if(trans(begin,words[i])){
                q.add(i);
                visit[i] = 1;
            }
        }
        
        while(!q.isEmpty()){
            int out = q.remove();
            if(words[out].equals(target)){
                answer = visit[out];
            }
            for(int i=0; i<words.length; i++){
                if(trans(words[out],words[i]) && visit[i]==0){
                    visit[i] = visit[out] + 1;
                    q.add(i);
                }
            }
        }
        return answer;
    }
    public boolean trans(String a, String b){
        int len = a.length();
        int cnt = 0;
        for(int i=0; i<len; i++){
            if(a.charAt(i) != b.charAt(i)){
                cnt++;
            }
        }
        return cnt == 1;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    n = len(words)
    visit = [0] * n
    q = []
    
    if target  not in words:
        key = -1
        return 0
    else:
        l = len(begin)
        q.append([begin,0])
        while q:
            x,dep = q.pop(0)
            answer = dep
            if x == target:
                break
            
            for i in range(n):
                cnt = 0
                if visit[i] == 0:
                    for j in range(l):
                        if words[i][j] == x[j]:
                            cnt += 1
                    if cnt == l-1:
                        q.append([words[i],dep+1])
                        visit[i] == 1
                    
        
        return answer
    